pkg/receive/handler.go

```diff
@@ -1195,35 +1227,139 @@ func (p *peerGroup) close(addr string) error {
 		return nil
 	}
 
+	p.cache[addr].turnOffGoroutines()
 	delete(p.cache, addr)
-	if err := c.Close(); err != nil {
+	if err := c.cc.Close(); err != nil {
 		return fmt.Errorf("closing connection for %s", addr)
 	}
 
 	return nil
 }
 
-func (p *peerGroup) get(ctx context.Context, addr string) (storepb.WriteableStoreClient, error) {
+type peerWorkResponse struct {
+	er  endpointReplica
+	err error
+}
+
+func (p *peerWorker) RemoteWriteAsync(ctx context.Context, req *storepb.WriteRequest, er endpointReplica, seriesIDs []int, responseWriter chan writeResponse, cb func()) {
+	p.initWorkers()
+
+	w := peerWorkItem{
+		cc:         p.cc,
+		req:        req,
+		workResult: make(chan peerWorkResponse, 1),
+		ctx:        ctx,
+		er:         er,
+
+		sendTime: time.Now(),
+	}
+
+	p.work <- w
+	res := <-w.workResult
+
+	responseWriter <- newWriteResponse(seriesIDs, res.err, er)
+	cb()
+}
+
+func (p *peerGroup) get(ctx context.Context, addr string) (WriteableStoreAsyncClient, error) {
 	// use a RLock first to prevent blocking if we don't need to.
 	p.m.RLock()
 	c, ok := p.cache[addr]
 	p.m.RUnlock()
 	if ok {
-		return storepb.NewWriteableStoreClient(c), nil
+		return c, nil
 	}
 
 	p.m.Lock()
 	defer p.m.Unlock()
 	// Make sure that another caller hasn't created the connection since obtaining the write lock.
 	c, ok = p.cache[addr]
 	if ok {
-		return storepb.NewWriteableStoreClient(c), nil
+		return c, nil
 	}
 	conn, err := p.dialer(ctx, addr, p.dialOpts...)
 	if err != nil {
 		return nil, errors.Wrap(err, "failed to dial peer")
 	}
 
-	p.cache[addr] = conn
-	return storepb.NewWriteableStoreClient(conn), nil
+	p.cache[addr] = newPeerWorker(conn, p.forwardDelay, p.asyncWorkerCount)
+	return p.cache[addr], nil
+}
+
+type peerWorker struct {
+	cc *grpc.ClientConn
+
+	work              chan peerWorkItem
+	turnOffGoroutines func()
+
+	initWorkersOnce  sync.Once
+	asyncWorkerCount uint
+	forwardDelay     prometheus.Histogram
+}
+
+func (pw *peerWorker) initWorkers() {
+	pw.initWorkersOnce.Do(func() {
+		work := make(chan peerWorkItem)
+		pw.work = work
+
+		ctx, cancel := context.WithCancel(context.Background())
+		pw.turnOffGoroutines = cancel
+
+		for i := 0; i < int(pw.asyncWorkerCount); i++ {
+			go func() {
+				for {
+					select {
+					case <-ctx.Done():
+						return
+					case w := <-work:
+						pw.forwardDelay.Observe(time.Since(w.sendTime).Seconds())
+
+						tracing.DoInSpan(w.ctx, "receive_forward", func(ctx context.Context) {
```

Comment:
`w.ctx` is only used here and not passed anywhere else. I don't see if being used for cancellation.

Reply:
`ctx` inside of the function inherits `w.ctx`:
```
							_, err := storepb.NewWriteableStoreClient(pw.cc).RemoteWrite(ctx, w.req)
```

Reply:
the `ctx` in the line you quote comes from `ctx, cancel := context.WithCancel(context.Background())` on L1305, no? It has nothing to do with `w.ctx`. 🤔

Reply:
Ahhh, that one from L1305 is being shadowed by the one in the anonymous function. Maybe we could use different names to avoid confusion?